src/lib/Microsoft.Health.Fhir.Ingest/Data/MeasurementToEventMessageAsyncCollector.cs

```diff
@@ -39,6 +47,38 @@ public async Task AddAsync(IMeasurement item, CancellationToken cancellationToke
             await _eventHubService.SendAsync(eventList, partitionKey, cancellationToken).ConfigureAwait(false);
         }
 
+        public async Task AddAsync(IEnumerable<IMeasurement> items, CancellationToken cancellationToken = default)
+        {
+            EnsureArg.IsNotNull(items, nameof(items));
+
+            using (var hasher = _hashCodeFactory.CreateDeterministicHashCodeGenerator())
+            {
+                var submissionTasks = items
+                .GroupBy(m =>
+                {
+                    var partitionKey = hasher.GenerateHashCode(m.DeviceId.ToLower()) % _partitionCount;
+                    var partitionKeyAsString = partitionKey.ToString(CultureInfo.InvariantCulture);
+                    return partitionKeyAsString;
+                })
+                .Select(async grp =>
+                {
+                    var partitionKey = grp.Key;
+                    var eventList = new List<EventData>();
+                    foreach (var m in grp)
+                    {
+                        var measurementContent = JsonConvert.SerializeObject(m, Formatting.None);
+                        var contentBytes = Encoding.UTF8.GetBytes(measurementContent);
+                        var eventData = new EventData(contentBytes);
+                        eventList.Add(eventData);
```

Comment:
>Add [](http://example.com/codeflow?start=34&length=3)

We should use EventDataBatch instead of a collection of events.  The EventDataBatch checks the total batch size client side and will error if the message added exceeds it.

We will want to modify the logic to catch this exception.  If we hit it we should flush immediately a then proceed with a new batch adding the event that triggered the error in the first place.

The underlying IEventHubMessageService already supports EventDataBatch with the signature SendAsync(EventDataBatch eventData, CancellationToken token)

Reply:
Good catch. Will do.